cprover_bindings/src/irep/goto_binary_serde.rs

```diff
@@ -0,0 +1,1462 @@
+        std::mem::size_of::<Self>()
+            + self.string_cache.dynamic_usage()
+            + self.inv_string_cache.dynamic_usage()
+            + self.cache.dynamic_usage()
+            + self.inv_cache.dynamic_usage()
+    }
+
+    fn dynamic_usage_bounds(&self) -> (usize, Option<usize>) {
+        let s = std::mem::size_of::<Self>();
+        let (l1, u1) = self.string_cache.dynamic_usage_bounds();
+        let (l2, u2) = self.inv_string_cache.dynamic_usage_bounds();
+        let (l3, u3) = self.cache.dynamic_usage_bounds();
+        let (l4, u4) = self.inv_cache.dynamic_usage_bounds();
+        let l = l1 + l2 + l3 + l4 + s;
+        let u = u1
+            .and_then(|u1| u2.and_then(|u2| u3.and_then(|u3| u4.map(|u4| u1 + u2 + u3 + u4 + s))));
+        (l, u)
+    }
+}
+
+#[cfg(test)]
+impl<'a, W> DynamicUsage for GotoBinarySerializer<'a, W>
+where
+    W: Write,
+{
+    fn dynamic_usage(&self) -> usize {
+        std::mem::size_of::<Self>()
+            + self.buf.dynamic_usage()
+            + self.numbering.dynamic_usage()
+            + self.irep_count.dynamic_usage()
+            + self.string_count.dynamic_usage()
+    }
+
+    fn dynamic_usage_bounds(&self) -> (usize, Option<usize>) {
+        let s = std::mem::size_of::<Self>();
+        let (l1, u1) = self.buf.dynamic_usage_bounds();
+        let (l2, u2) = self.numbering.dynamic_usage_bounds();
+        let (l3, u3) = self.irep_count.dynamic_usage_bounds();
+        let (l4, u4) = self.string_count.dynamic_usage_bounds();
+        let l = l1 + l2 + l3 + l4 + s;
+        let u = u1
+            .and_then(|u1| u2.and_then(|u2| u3.and_then(|u3| u4.map(|u4| u1 + u2 + u3 + u4 + s))));
+        (l, u)
+    }
+}
+
+#[cfg(test)]
+impl<R> DynamicUsage for GotoBinaryDeserializer<R>
+where
+    R: Read,
+{
+    fn dynamic_usage(&self) -> usize {
+        std::mem::size_of::<Self>()
+            + self.numbering.dynamic_usage()
+            + self.irep_count.dynamic_usage()
+            + self.irep_map.dynamic_usage()
+            + self.string_count.dynamic_usage()
+            + self.string_map.dynamic_usage()
+    }
+
+    fn dynamic_usage_bounds(&self) -> (usize, Option<usize>) {
+        let s = std::mem::size_of::<Self>();
+        let (l1, u1) = self.numbering.dynamic_usage_bounds();
+        let (l2, u2) = self.irep_count.dynamic_usage_bounds();
+        let (l3, u3) = self.irep_map.dynamic_usage_bounds();
+        let (l4, u4) = self.string_count.dynamic_usage_bounds();
+        let (l5, u5) = self.string_map.dynamic_usage_bounds();
+        let l = l1 + l2 + l3 + l4 + l5 + s;
+        let u = u1.and_then(|u1| {
+            u2.and_then(|u2| {
+                u3.and_then(|u3| u4.and_then(|u4| u5.map(|u5| u1 + u2 + u3 + u4 + u5 + s)))
+            })
+        });
+        (l, u)
+    }
+}
+
+#[cfg(test)]
+#[derive(Debug)]
+/// Structural sharing statistics
+struct SharingStats {
+    // Number of structurally unique objects
+    _nof_unique: usize,
```

Comment:
Why are you using `_` in the name? In rust, leading _ means it's unused

Reply:
that's why I use it. When all you do is print the field Rustc tells you the field is not used

Reply:
Oh... you are only using them as part of debug. Ok, we can live with that for now.